### PR TITLE
Update to .NET SDK 7.0 Preview 2

### PR DESCRIPTION
**Pull request by bricelam**
~~Still waiting for the official release, so the version number may change again before I merge, but this one is close.~~

### CODE REVIEW

**Comment by bricelam** [comment]
@dotnet/efteam Should we update our tests to run on net7.0 or wait?

**Comment by bricelam** [comment]
Looks like there are a few failures in the analyzers tests.

> The collection was expected to contain a single element

File: test/EFCore.Analyzers.Tests/UninitializedDbSetDiagnosticSuppressorTest.cs [hunk 1]
```diff
@@ -146,8 +146,10 @@ public class Blog
         Assert.False(diagnostic.IsSuppressed);
     }
 
-    protected Task<Diagnostic[]> GetDiagnosticsFullSourceAsync(string source)
-        => base.GetDiagnosticsFullSourceAsync(source, analyzerDiagnosticsOnly: false);
+    protected async Task<Diagnostic[]> GetDiagnosticsFullSourceAsync(string source)
+        => (await base.GetDiagnosticsFullSourceAsync(source, analyzerDiagnosticsOnly: false))
+            .Where(d => d.Id != "CS1701") // Assuming assembly reference '...Version=6.0.0...' matches '...Version=7.0.0...
```
Comment (roji) [inline test/EFCore.Analyzers.Tests/UninitializedDbSetDiagnosticSuppressorTest.cs:151]: Thanks... I'll maybe look at this again when I finish the work on the suppressor...

**Review by roji — APPROVED** [review]

File: test/EFCore.Analyzers.Tests/UninitializedDbSetDiagnosticSuppressorTest.cs (hunk 1 above)
Comment (bricelam) [reply test/EFCore.Analyzers.Tests/UninitializedDbSetDiagnosticSuppressorTest.cs:151]: Note that this and 1702 are disable by default in all C# projects:

https://github.com/dotnet/sdk/blob/f2cc77b60cdc2a4e8b393adf7b855b478ad79671/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.CSharp.props#L16

**Comment by roji** [comment]
Huh? Tried to fix the trimming by running the resulting executable from the right directory (net7.0 instead of net6.0), now there's a new test failure on AddressAndPayment_RedirectToCompleteWhenSuccessful :interrobang:

**Comment by bricelam** [comment]
@maumar, do you want to dig into the MusicStoreInMemoryTest.AddressAndPayment_RedirectToCompleteWhenSuccessful failure?

    Assert.Equal() Failure
    Expected: 0
    Actual:   (null)

**Comment by bricelam** [comment]
And now there's new exceptions being thrown from System.Text.Json.
```
TypeLoadException : Could not load type 'System.Runtime.CompilerServices.Unsafe' from assembly 'System.Runtime, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
   at System.Text.Json.JsonReaderHelper.IndexOfOrLessThan(Byte& searchSpace, Byte value0, Byte value1, Byte lessThan, Int32 length)
   at System.Text.Json.Utf8JsonReader.ConsumeString()
   at System.Text.Json.Utf8JsonReader.ConsumePropertyName()
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at Microsoft.Extensions.DependencyModel.DependencyContextJsonReader.Read(Utf8JsonReader reader)
   at Microsoft.Extensions.DependencyModel.DependencyContextJsonReader.Read(Stream stream)
   at Microsoft.Extensions.DependencyModel.DependencyContextLoader.LoadAssemblyContext(Assembly assembly, IDependencyContextReader reader)
   at Microsoft.Extensions.DependencyModel.DependencyContextLoader.Load(Assembly assembly)
   at Microsoft.Extensions.DependencyModel.DependencyContext.Load(Assembly assembly)
```

**Comment by bricelam** [comment]
@Pilchie Is the Json exception a known issue?

**Comment by Pilchie** [comment]
Not known to me.  @danmoseley - any ideas?

**Comment by Pilchie** [comment]
@dotnet/aspnet-build - any of you seen this in .NET 7 builds?

**Comment by wtgodbe** [comment]
Looks like you might be building one version of the runtime while running against another - there may be other `net6.0` TFMs to update. I see it's referenced in 51 files but this PR only touches 41: https://github.com/dotnet/efcore/search?q=net6.0

**Comment by ajcvickers** [comment]
@wtgodbe We still want to target .NET 6 with our product assemblies, but we want our tests to use .NET 7. (Ideally, we won't target .NET 7 even when we ship RTM since that forces people off an LTS framework, which many people don't want to do.)

**Comment by Pilchie** [comment]
Just saw https://github.com/dotnet/aspnetcore/pull/40471 from @jkotas which looks like it might be related?

**Comment by jkotas** [comment]
> Just saw https://github.com/dotnet/aspnetcore/pull/40471 from @jkotas which looks like it might be related?

Yes, it is related, but I do not think it is the root cause of the build break. The change that removed S.R.CS.Unsafe package in dotnet/runtime is not included in .NET 7 preview1. If you see build breaks due to do this package removal, you are mixing .NET 7 preview1 bits with some newer bits. It is not unusual to see intra-release breaking changes in previews like this one. Mixing bits from different previews will expose you to the intra-release breaking changes like that.

**Comment by bricelam** [comment]
Thanks! We typically only hit a few of these breaks each release. We find that saying on released previews of the SDK and runtime helps contributors immensely since they don't have to worry about setting environment variables before launching their IDE.

I'll pin a few packages to their Preview 1 versions for now. If this continues to be a problem, we can consider [flowing the runtime](https://github.com/dotnet/efcore/pull/25029) again (at the expense of productivity and contributors).

**Comment by bricelam** [comment]
@maumar Looks like SQL Server and SQLite are now failing with the same error.

**Comment by maumar** [comment]
@bricelam AddressAndPayment_RedirectToCompleteWhenSuccessful  also happens due to 

```
'Could not load type 'System.Runtime.CompilerServices.Unsafe' from assembly 'System.Runtime, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.'
```

**Comment by maumar** [comment]
@bricelam 
```
Could not load type 'System.Runtime.CompilerServices.Unsafe' from assembly 'System.Runtime, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.":"System.Runtime.CompilerServices.Unsafe"

   at Microsoft.Extensions.Primitives.StringValues.get_Count()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.MusicStoreTestBase`1.CheckoutController.<AddressAndPayment>d__3.MoveNext() in 
```

**Comment by bricelam** [comment]
New failures in EFCore.OData.FunctionalTests after updating to Preview 2.

**Comment by smitpatel** [comment]
```
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.IO.IOException : The response ended prematurely.
```

**Comment by bricelam** [comment]
Yep. That's the one.

@maumar Would it be possible (without a ton of infrastructure) to show the server-side errors on failure?

**Comment by maumar** [comment]
@bricelam OData tests fail due to:

```
Could not load type 'System.Runtime.CompilerServices.Unsafe' from assembly 'System.Runtime, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.EnsureHostHeaderExists()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
Exception thrown: 'System.Net.Http.HttpRequestException' in System.Private.CoreLib.dll
```

File: eng/Versions.props [hunk 2]
```diff
@@ -15,17 +15,17 @@
...
   </PropertyGroup>
   <PropertyGroup Label="Dependencies from dotnet/runtime">
-    <SystemCollectionsImmutableVersion>7.0.0-preview.3.22171.1</SystemCollectionsImmutableVersion>
-    <SystemDiagnosticsDiagnosticSourceVersion>7.0.0-preview.3.22171.1</SystemDiagnosticsDiagnosticSourceVersion>
-    <MicrosoftExtensionsCachingMemoryVersion>7.0.0-preview.3.22171.1</MicrosoftExtensionsCachingMemoryVersion>
-    <MicrosoftExtensionsConfigurationVersion>7.0.0-preview.3.22171.1</MicrosoftExtensionsConfigurationVersion>
-    <MicrosoftExtensionsConfigurationAbstractionsVersion>7.0.0-preview.3.22171.1</MicrosoftExtensionsConfigurationAbstractionsVersion>
-    <MicrosoftExtensionsConfigurationEnvironmentVariablesVersion>7.0.0-preview.3.22171.1</MicrosoftExtensionsConfigurationEnvironmentVariablesVersion>
-    <MicrosoftExtensionsConfigurationJsonVersion>7.0.0-preview.3.22171.1</MicrosoftExtensionsConfigurationJsonVersion>
+    <SystemCollectionsImmutableVersion>7.0.0-preview.2.22152.2</SystemCollectionsImmutableVersion>
```
Comment (smitpatel) [inline eng/Versions.props:18]: With this change we have downgraded our dependencies in main branch. Is that required? Do we need to get main to upgrade dependencies again?
cc: @dotnet/aspnet-build

Comment (bricelam) [reply eng/Versions.props:18]: Will unpin once we upgrade to Preview 3 of the SDK (see conversation above)

**Pull request merged**